### PR TITLE
Create separate commands for embedding, clustering, and distance matrix creation

### PR DESCRIPTION
**Pull request by nandsra21**
Closes https://github.com/blab/cartography/issues/33

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -158,3 +158,5 @@ cython_debug/
 #  and can be added to the global gitignore or merged into this file.  For a more nuclear
 #  option (not recommended) you can uncomment the following to ignore the entire idea folder.
 #.idea/
+test/
+pathogenEmbed.yml
```
Comment (huddlej) [inline .gitignore:162]: Why do we need to ignore these specific paths?

File: setup.py [hunk 2]
```diff
@@ -42,7 +42,9 @@
     ],
     entry_points = {
         "console_scripts": [
-            "embed = embed.__main__:main",
+            "pathogen-embed = pathogen_embed.__main__:main",
+            "pathogen-distance = pathogen_distance.__main__:main",
+            "pathogen-cluster = pathogen_cluster.__main__:main"
```
Comment (huddlej) [inline setup.py:47]: Since this interface allows us to connect specific command line scripts to specific functions in a given `__main__` module, we could place all of the code for these different commands in the same module under different function names and refer to those here. For example:

```python
            "pathogen-embed = pathogen_embed.__main__:embed",
            "pathogen-distance = pathogen_embed.__main__:distance",
            "pathogen-cluster = pathogen_embed.__main__:cluster"
```

I think this approach would allow us to reference shared code from different functions without duplicating that code across multiple modules.

File: src/pathogen_embed/__init__.py~
Comment (huddlej) [inline src/pathogen_embed/__init__.py~:1]: We should drop this file, since it is a duplicate with a `~` suffix.

File: CHANGES.md [hunk 3]
```diff
@@ -1,4 +1,7 @@
 # CHANGELOG
+## 1.0.0
+* Creating separate commands for embedding, clustering, and distance matrix creation (pathogen-embed, pathogen-cluster, pathogen-distance) ([28aa891](https://github.com/blab/pathogen-embed/commit/28aa8912a24138a3dcd54fa43f90a5e10b4aa86e))
```
Comment (huddlej) [inline CHANGES.md:3]: For the change logs in Nextstrain, we tend to reference the PR that introduces the change. This allows us to refer to multiple commits with a single link. For an example, [see Augur's change log](https://raw.githubusercontent.com/nextstrain/augur/master/CHANGES.md). I slightly prefer the direct Markdown link approach you use here over the reference-based approach in the Augur change log, but you can use whichever style you prefer as long as it is consistent throughout the log.

File: src/pathogen_cluster/__init__.py [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+__credits__ = 'Bedford Lab, Vaccine and Infectious Disease Division, Fred Hutchinson Cancer Research Center, Seattle, WA, USA'
+
+import argparse
+import sys
+from .pathogen_cluster import pathogen_cluster
+
+def make_parser():
+    parser = argparse.ArgumentParser(description = "Reduced dimension embeddings for pathogen sequences", formatter_class=argparse.ArgumentDefaultsHelpFormatter)
+
+    # correct
```
Comment (huddlej) [inline src/pathogen_cluster/__init__.py:17]: What does this comment refer to? Is it a TODO?

File: src/pathogen_cluster/__init__.py [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+__version__ = "1.0.0"
+__author__ = 'Sravani Nanduri, John Huddleston'
+__credits__ = 'Bedford Lab, Vaccine and Infectious Disease Division, Fred Hutchinson Cancer Research Center, Seattle, WA, USA'
+
+import argparse
+import sys
+from .pathogen_cluster import pathogen_cluster
+
+def make_parser():
+    parser = argparse.ArgumentParser(description = "Reduced dimension embeddings for pathogen sequences", formatter_class=argparse.ArgumentDefaultsHelpFormatter)
```
Comment (huddlej) [inline src/pathogen_cluster/__init__.py:15]: We'll need to update this description to reflect what the cluster command does.

File: src/pathogen_cluster/__init__.py [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+
+    # correct
+    parser.add_argument("--embedding", help="The embedding to assign clustering labels to via HDBSCAN (https://hdbscan.readthedocs.io/en/latest/how_hdbscan_works.html)")
+    parser.add_argument("--random-seed", default = 314159, type=int, help="an integer used for reproducible results.")
+    parser.add_argument("--min-size", type=int, default=5, help="minimum cluster size for HDBSCAN")
+    parser.add_argument("--min-samples", type=int, default=5, help="minimum number of sample to seed a cluster for HDBSCAN. Lowering this value reduces number of samples that do not get clustered.")
+    parser.add_argument("--distance-threshold", type=float, help="The float value for the distance threshold by which to cluster data in the embedding and assign labels via HDBSCAN. If no value is given in cluster-data or cluster-threshold, the default distance threshold of 0.0 will be used.")
+    parser.add_argument("--output-dataframe", help="a csv file outputting the embedding with the strain name and its components.")
```
Comment (huddlej) [inline src/pathogen_cluster/__init__.py:23]: We should probably make this data frame output and the embedding input required arguments.

File: src/pathogen_cluster/pathogen_cluster.py [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+    embedding_df = pd.read_csv(args.embedding, index_col=0)
+
+    clustering_parameters = {
+        **({"min_cluster_size": args.min_size} if args.min_size is not None else {}),
+        **({"min_samples": args.min_samples} if args.min_samples is not None else {}),
+        **({"cluster_selection_epsilon": args.distance_threshold} if args.distance_threshold is not None else {})
+    }
+
+    clusterer = hdbscan.HDBSCAN(**clustering_parameters)
+    clusterer_default = hdbscan.HDBSCAN()
```
Comment (huddlej) [inline src/pathogen_cluster/pathogen_cluster.py:18]: I think this "default" clusterer and its outputs were useful for our initial exploration of HDBSCAN, but now that we will have a separate cluster command that makes it easy to cluster with different settings, we can safely drop this default clusterer from this code.

File: src/pathogen_cluster/pathogen_cluster.py [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+
+    clusterer.fit(embedding_df)
+    clusterer_default.fit(embedding_df)
+    embedding_df["label"] = clusterer.labels_.astype(str)
```
Comment (huddlej) [inline src/pathogen_cluster/pathogen_cluster.py:22]: We'll want a way to set different names for these cluster labels, so we can run multiple cluster commands on the same embedding file and merge the final results without conflicts. We might want to add a command line parameter like `--label-attribute` to support this functionality.

File: src/pathogen_embed/__init__.py
Comment (huddlej) [inline src/pathogen_embed/__init__.py:1]: I like the renaming of the source directory here from `embed` to `pathogen_embed`.

File: src/pathogen_embed/__init__.py [hunk 9]
```diff
@@ -4,26 +4,22 @@
...
     parser.add_argument("--distance-matrix", help="a distance matrix that can be read in by pandas, index column as row 0")
     parser.add_argument("--separator", default=",", help="separator between columns in the given distance matrix")
-    parser.add_argument("--alignment", help="an aligned FASTA file to create a distance matrix with. Make sure the strain order in this file matches the order in the distance matrix.")
-    parser.add_argument("--cluster-data", help="The file (same separator as distance-matrix) that contains the distance threshold by which to cluster data in the embedding and assign labels given via HDBSCAN (https://hdbscan.readthedocs.io/en/latest/how_hdbscan_works.html). If no value is given in cluster-data or cluster-threshold, the default distance threshold of 0.0 will be used. If any additional columns in this file match embedding-specific parameters (e.g., 'n_components' or 'learning_rate'), the values from the first record of this file will override default values or values provided by the command line arguments.")
```
Comment (huddlej) [inline src/pathogen_embed/__init__.py:21]: We need to keep an argument like this `--cluster-data` argument to allow users (and our own workflows) to pass through a file with embedding method parameters that override the defaults. Since this command no longer performs clustering, we can rename the argument to something more accurate like `--embedding-parameters`.

File: src/pathogen_embed/pathogen_embed.py [hunk 10]
```diff
@@ -296,18 +259,6 @@ def embed(args):
...
-    if cluster_data is not None and args.command != "pca":
-        for key, value in cluster_data.items():
-            if key in embedding_parameters:
-                value_type = type(embedding_parameters[key])
-                print(
-                    f"INFO: Replacing embedding parameter {key} value of '{embedding_parameters[key]}' with '{value_type(value)}' provided by '{args.cluster_data}'.",
-                    file=sys.stderr
-                )
-                embedding_parameters[key] = value_type(value)
-
```
Comment (huddlej) [inline src/pathogen_embed/pathogen_embed.py:310]: Related to the comment above about keeping a `--cluster-data`-style argument for this command, we need to keep this block of logic, so we can override default parameters with a single file from the command line.

**Review by huddlej — CHANGES_REQUESTED** [review]
Thank you for setting this up, @nandsra21! The biggest change we should look into is how to put all of the logic for these separate commands into the same module, so we can avoid code duplication. Consolidating into one module should also minimize duplication of other boilerplate code like most of the `__main__.py` code.

Other major changes are mentioned below including:

 - keeping a command line argument and logic to override embedding parameters with a file input
 - allowing users to specify the name of the cluster label attribute, so we can run clustering multiple times on the same inputs

We can chat about this all more today at our meeting, though, so no worries about responding to any of these right away.

File: src/pathogen_distance/__init__.py [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+__version__ = "1.0.0"
+__author__ = 'Sravani Nanduri, John Huddleston'
+__credits__ = 'Bedford Lab, Vaccine and Infectious Disease Division, Fred Hutchinson Cancer Research Center, Seattle, WA, USA'
+
+import argparse
+import sys
+from .pathogen_distance import pathogen_distance
+
+def make_parser():
+    parser = argparse.ArgumentParser(description = "Reduced dimension embeddings for pathogen sequences", formatter_class=argparse.ArgumentDefaultsHelpFormatter)
```
Comment (huddlej) [inline src/pathogen_distance/__init__.py:15]: We need to update this description, too.

File: .gitignore (hunk 1 above)
Comment (nandsra21) [reply .gitignore:162]: I have a "test" repo that is filled with fasta files and stuff I used for testing the commands - didn't think it would make sense to push here! And the yml file is just not working yet so I thought I'd add it when it works as expected.

File: src/pathogen_cluster/__init__.py (hunk 4 above)
Comment (nandsra21) [reply src/pathogen_cluster/__init__.py:17]: I should definitely delete this - a relic of me updating the command line arguments 😅

File: src/pathogen_embed/pathogen_embed.py [hunk 12]
```diff
@@ -346,10 +318,39 @@ def embed(args):
...
+    clustering_parameters = {
+        **({"min_cluster_size": args.min_size} if args.min_size is not None else {}),
+        **({"min_samples": args.min_samples} if args.min_samples is not None else {}),
+        **({"cluster_selection_epsilon": args.distance_threshold} if args.distance_threshold is not None else {})
+    }
+
+    clusterer = hdbscan.HDBSCAN(**clustering_parameters)
+
+    clusterer.fit(embedding_df)
+    embedding_df[f"label_{args.label_attribute}"] = clusterer.labels_.astype(str)
```
[the remark this answers is not in the document]
Comment (huddlej) [reply src/pathogen_embed/pathogen_embed.py]: I made this change in [66f5204](https://github.com/blab/pathogen-embed/pull/2/commits/66f52049e074a9c8696d19e3547ce6af8770c164).

File: src/pathogen_embed/pathogen_embed.py [hunk 13]
```diff
@@ -313,30 +292,23 @@ def embed(args):
...
+        embedding_df.columns=[args.command + str(i) for i in range(1, args.components + 1)]
     else:
         embedding_df.columns = [args.command.replace('-', '') + "_x" , args.command.replace('-', '') + "_y"]
 
     if args.command == "pca":
 
         #add explained variance as the first row of the dataframe
         explained_variance = pd.DataFrame([round(pca.explained_variance_ratio_[i],4) for i in range(0,len(pca.explained_variance_ratio_))], columns=["explained variance"])
-        explained_variance["principal components"] = [i for i in range(1, n_components + 1)]
+        explained_variance["principal components"] = [i for i in range(1, args.components + 1)]
```
[the remark this answers is not in the document]
Comment (huddlej) [reply src/pathogen_embed/pathogen_embed.py]: Fixed in [a1b5a95](https://github.com/blab/pathogen-embed/pull/2/commits/a1b5a956e78be508347a7b6d8bfedd82d47b0b25).

File: src/pathogen_embed/pathogen_embed.py [hunk 14]
```diff
@@ -313,30 +292,23 @@ def embed(args):
...
         # Output Embedding
-            # create dictionary to be "wrapped" by write_json
+        # create dictionary to be "wrapped" by write_json
 
         embedding_df = pd.DataFrame(embedding)
         embedding_df.index = list(distance_matrix.index)
 
     if args.command == "mds" or args.command == "pca":
-        embedding_df.columns=[args.command + str(i) for i in range(1, n_components + 1)]
+        embedding_df.columns=[args.command + str(i) for i in range(1, args.components + 1)]
```
[the remark this answers is not in the document]
Comment (huddlej) [reply src/pathogen_embed/pathogen_embed.py]: Fixed in [a1b5a95](https://github.com/blab/pathogen-embed/pull/2/commits/a1b5a956e78be508347a7b6d8bfedd82d47b0b25).

File: src/pathogen_embed/pathogen_embed.py [hunk 15]
```diff
@@ -258,8 +236,9 @@ def embed(args):
         genomes_df = pd.DataFrame(numbers)
         genomes_df.columns = ["Site " + str(k) for k in range(0,len(numbers[i]))]
 
+
         #performing PCA on my pandas dataframe
-        pca = PCA(n_components=n_components, svd_solver='full') #can specify n, since with no prior knowledge, I use None
+        pca = PCA(n_components=args.components, svd_solver='full')
```
[the remark this answers is not in the document]
Comment (huddlej) [reply src/pathogen_embed/pathogen_embed.py]: Fixed in [a1b5a95](https://github.com/blab/pathogen-embed/pull/2/commits/a1b5a956e78be508347a7b6d8bfedd82d47b0b25).

**Review by huddlej — COMMENTED** [review]
Thanks for all the quick updates, @nandsra21. I made some comments below, but then out of interest I ended up implementing some of them.

After reading more about the Numba warnings issue and getting annoyed with it myself as a user of our tool, I ended up ignoring these warnings in 00af5f768f90b80f6adb8da32dd0bc78b7b86245.

I also fixed a bug in caused by my poor recommendation of a variable name for the external embedding parameters. See db224208aad048f8a1be3680d57ea6f2318b0fe3 for the diff and details in the commit message.

Although this is looking really good, I want to test the functionality more, to make sure nothing got broken in all the moving code. I'll post back here after spending more time with the code tomorrow.
